Add modular planner interface.

ibrahiminfinite opened the pull request:
This PR adds a new planner interface that makes it easy to integrate and use new planners in pyrobosim.
The new interface does not break any existing pipelines as the APIs and behaviours remain the same.

Will close #81, #15


TODO :

- [x] Add planner interface
- [x] Planner base class
- [x] A* planner implementation (grid)
- [x] A* planner implementation (graph)
- [x] RRT implementation (graph)
- [x] Fix RRT connection issue
- [x] PRM implementation (graph)
- [x] Add method to print metrics
- [x] Support displaying the planner graph 
- [x] Add Waypoint reduction for planners 
- [x] Add demo for new planner interface
- [x] Update YAML loader to use new planner interface. (To be implemented after API unification PR is merged)
- [x] Refactor search_graph and cascading changes.
- [x] Add documentation for how to use the interface to implement new planners.

Edit :  Add `graph_from_world` method (No planners currently need this, postponing till this is needed)

NOTE:  Might break the pipeline when loaded as the "world" global planner.  However I am not  sure if the world should have a planner as the "planning" is a behaviour of the robot and so, it makes sense for only robots to have planners.  Might be a good idea for a robot to have both a global and local planner , instead of the global planner being a part of the world.

ibrahiminfinite commented:
> However I am not sure if the world should have a planner as the "planning" is a behaviour of the robot and so, it makes sense for only robots to have planners. Might be a good idea for a robot to have both a global and local planner , instead of the global planner being a part of the world.

@sea-bass , I would like to know your opinion on this.

sea-bass commented:
I think separating planners from the world is the way to go.

At some point I thought, why not just have 1 planner that multiple robots can call? But then if they're different sizes or have different dynamics, that doesn't reallt work.

Also, this tool's original definition of "global" vs "local" is different from what it typically means. All the planners available now are technically global planners, in that they use knowledge about the world to plan the whole path.

So go with your idea!

sea-bass commented:
The above all seems reasonable -- thanks!

ibrahiminfinite commented:
There are still stuff left to do here, but I think it is ready for a preliminary review.

sea-bass commented on pyrobosim/pyrobosim/data/test_world.yaml:
```diff
@@ -22,12 +22,12 @@ robots:
-    #   bidirectional: true
-    #   rrt_star: true
-    #   rewire_radius: 1.0
+    path_planner:
+      type: rrt
+      max_connection_dist: 0.25
+      bidirectional: true
+      rrt_star: false
+      rewire_radius: 1.0
```
This is exactly how I envisioned it would become, so looks good!

sea-bass commented on pyrobosim/pyrobosim/core/world.py:
```diff
@@ -537,7 +511,7 @@ def add_object(self, **object_config):
             if not is_valid_pose:
                 warnings.warn(
-                    f"Object {obj.name} in collision or not in location {loc.name}. Cannot add to world."
+                    f"Object {obj.name} in collision or not in location {obj_spawn.name}. Cannot add to world."
```
nice catch, thanks

sea-bass reviewed (commented):
I think this is looking great. Haven't yet done testing until you tell me it's ready for that.

Here are my high level thoughts:
* As you pointed out, we need some way to read the occupancy grid from YAML. And I think in accordance with your own suggestions, we could simply make an `OccupancyGrid.fromYaml()` method. (Actually, the same is needed for the pose class because right now we specify it as a list which is prone to error)
* I like that common functionality like `plot()` was consolidated, but please remember that for the PRM and RRT there are also functions to show the graphs themselves. Maybe this could also be something pulled into the generic function, where you can optionally pass it a graph to plot, or else you will need to overload it.
* As you test things out, I wouldn't mind more unit tests popping up here and there :)

sea-bass commented on pyrobosim/pyrobosim/navigation/planner_base.py:
```diff
@@ -0,0 +1,88 @@
+        f = plt.figure()
+        ax = f.add_subplot(111)
+        self.plot(ax)
+        plt.title("A*")
```
This won't always be A*, of course.

Should either be passed in as an argument or read in from e.g. the config. For example, if there was a `name` attribute in the dictionary, it could use that; else maybe it just says "Path" as the title or something

ibrahiminfinite replied:
We can use the planner name there like how it is done inside the `info()` method.

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/path_planner.py:
```diff
@@ -0,0 +1,86 @@
+        :param planner_config: The configuration to be used with the specified planner type.
+        :type planner_config: dict
+
+        The planner_config can have a many options within it.
+
+        1. grid - None by default, but can be assigned to an OccupancyGrid object.
+                  If set to an OccupancyGrid, the planner will use the grid for planning.
+                  If None, the planner will use the polygon based methods for planning.
+
+        These parameters apply to both RRT and PRM planners.
+
+        2. max_nodes_sampled (int): Specifies the maximum number of nodes to be generated for sampling-
+                  based planners.
+
+        3. max_connection_dist (float) : Specifies the maximum distance allowed between two nodes, in meters.
+
+        These parameters are only applicable to RRT planners
+
+        4. rrt_bidirectional (bool) : If true, will use bidirectional RRT.
+
+        5. rrt_connect (bool) :  If true, will use rrt connect.
+
+        6. rrt_star (bool) : If true, will use use rrt star for optimization.
+
+        7. rrt_rewire_radius (float) : The radius within which a node can find a rewire candidate, in meters.
```
I am not yet sure about if this should be here or moved out into their respective planner classes.

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/planner_base.py:
```diff
@@ -0,0 +1,88 @@
+        )
+
+    def info(self):
+        """Prints the information about the planning process"""
```
Yes, planners that need to show more info can override.

ibrahiminfinite commented on pyrobosim/pyrobosim/core/yaml_utils.py:
```diff
@@ -127,90 +126,37 @@ def add_robots(self):
-                rewire_radius=rewire_radius,
-            )
+            planner_config = {
+                "grid": None,  # TODO : How should a grid be added based on yaml config
```
I dont know if we need a `OccupancyGrid.fromYaml()`.
Since the yaml info will be read anyway when loading the robot's planner, we can just call the `occupancy_grid_from_world()` there and pass in the grid to the planner.

ibrahiminfinite commented:
The graphs can be displayed now. 
![rrt_graph](https://user-images.githubusercontent.com/12377945/229334342-5d53309c-de38-486b-9f1e-ea3afdb43a1a.png)
![prm_graph](https://user-images.githubusercontent.com/12377945/229334349-fe470431-7efb-49ea-9ccf-ef513e9b2121.png)

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/planner_base.py:
```diff
@@ -12,6 +12,7 @@ def __init__(self):
         self.start = None
         self.impl = None
         self.planning_time = 0
+        self.graphs = set()
```
Needs to be changed to `list`

ibrahiminfinite commented:
@sea-bass 
Only the documentation part remains now (tutorials and any docstring updates)

The issue with RRT* + Bidirectional seems to be fixed as well. Did not encounter it especially with `compress_path = True`.
Let me know if you encounter the issue in your testing.

sea-bass commented on pyrobosim/examples/demo.py:
```diff
@@ -122,14 +122,16 @@ def create_world(multirobot=False):
-        robby.set_path_planner(
-            RRTPlanner(world, bidirectional=True, rrt_connect=False, rrt_star=True)
-        )
+        planner_config = {
+            "grid": None,
+            "world": world,
+            "bidirectional": False,
+            "rrt_connect": False,
+            "rrt_star": False,
+        }
+        rrt = PathPlanner("rrt", planner_config)
+        robby.set_path_planner(rrt)
```
This is only adding a planner in the multirobot case. When I ran the demo without any extra arguments, I got a warning that the first robot (simply named "robot"), doesn't have a planner.

I guess a byproduct of this work is that if you instantiate a robot by default, it doesn't have a planner. So basically none of the demos actually work in their current state. Please check this.

When I did run this file with the `--world-file test_world.yaml` argument, it works but for some reason navigating to `counter0` kept failing. Also check this.

sea-bass commented on pyrobosim/examples/demo.py:
```diff
@@ -122,14 +122,16 @@ def create_world(multirobot=False):
-        robby.set_path_planner(
-            RRTPlanner(world, bidirectional=True, rrt_connect=False, rrt_star=True)
-        )
+        planner_config = {
+            "grid": None,
+            "world": world,
+            "bidirectional": False,
+            "rrt_connect": False,
+            "rrt_star": False,
+        }
+        rrt = PathPlanner("rrt", planner_config)
```
As discussed offline, I like the `**kwargs` idea where the lowest-level planner actually has named arguments. It would change this to:

```python
rrt = PathPlanner("rrt", **planner_config)
```

sea-bass commented on pyrobosim/examples/demo_astar.py:
```diff
@@ -0,0 +1,38 @@
+import os
```
This demo doesn't work because of circular imports. I think I broke some of the ordering in imports from the last PR, where now import order matters.

```
  File "/home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/core/robot.py", line 11, in <module>
    from ..utils.knowledge import resolve_to_object
ImportError: cannot import name 'resolve_to_object' from partially initialized module 'pyrobosim.utils.knowledge' (most likely due to a circular import) (/home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/utils/knowledge.py)
```

sea-bass commented on pyrobosim/examples/demo_prm.py:
```diff
@@ -15,15 +15,21 @@
 def test_prm():
     """Creates a PRM planner and plans"""
-    prm = PRMPlanner(world, max_nodes=100, max_connection_dist=1.5)
+    planner_config = {
```
This demo works, but the graph colors and line styles changed since `main`, and in my opinion they look a lot worse and harder to visualize. Consider using the same "gentler" semi-transparent blue as original.

Also, I'm not sure what's happening but this demo runs very slowly.

ibrahiminfinite replied:
The slowness is probably due to the the call to `show_path_and_planner()` rather than just `show_path()`.
Since it redraws the graph every time navigate is clicked.  I was supposed to change that but forgot. Will change it.

sea-bass commented on pyrobosim/examples/demo_rrt.py:
```diff
@@ -15,15 +15,22 @@
 def test_rrt():
     """Creates an RRT planner and plans"""
-    rrt = RRTPlanner(world, bidirectional=True, rrt_connect=False, rrt_star=True)
+    planner_config = {
```
Works, but the default `compress_path` option makes this look weird. I think it should be off by default, whereas I see that now this option is constantly hard-coded as true.

Furthermore, I would offer the suggestion to remove `compress_path` from all the planners, and instead offer it as a post-processing utility that people can opt into. Again, defaulting to false because I want to make sure that people see the "original" planner output by default for educational reasons.

ibrahiminfinite replied:
This is actually not hard-coded.
If you pass "compress_path: False" in the config, it will no longer compress the path.

sea-bass replied:
Oh, right, that just has the defaults. My bad. Well, then just changing the default to True will work then :)

sea-bass commented on test/planning/test_pddlstream_manip.py:
```diff
@@ -76,8 +76,15 @@ def create_test_world(add_alt_desk=True):
     # Create a search graph and motion planner
-    world.create_search_graph(max_edge_dist=3.0, collision_check_dist=0.05)
-    rrt = RRTPlanner(world, bidirectional=True, rrt_star=True)
+    # world.create_search_graph(max_edge_dist=3.0, collision_check_dist=0.05)
```
Remove this commented out line?

sea-bass reviewed (commented):
Gave it a test. Some high level comments:

* Please don't hard-code the "compress_path" option as true. Make it an option and set it false by default, so people get the "traditional" planner output by default.
* The regular demos currently don't work unless you load from file, because `Robot` objects don't have a default planner anymore (there is no more global planner)
* The `pyrobosim_ros` demos just do not work at the moment:

```
[demo.py-1] [INFO] [1680443536.118004148] [pyrobosim.demo]: World node started.
[demo.py-1] [INFO] [1680443536.118579273] [pyrobosim.demo]: Creating demo world programmatically.
[demo.py-1] Traceback (most recent call last):
[demo.py-1]   File "/home/sebastian/personal/pyrobosim_ws/install/pyrobosim_ros/lib/pyrobosim_ros/demo.py", line 120, in <module>
[demo.py-1]     node = create_ros_node()
[demo.py-1]   File "/home/sebastian/personal/pyrobosim_ws/install/pyrobosim_ros/lib/pyrobosim_ros/demo.py", line 109, in create_ros_node
[demo.py-1]     world = create_world()
[demo.py-1]   File "/home/sebastian/personal/pyrobosim_ws/install/pyrobosim_ros/lib/pyrobosim_ros/demo.py", line 88, in create_world
[demo.py-1]     world.create_search_graph(
[demo.py-1] AttributeError: 'World' object has no attribute 'create_search_graph'
```

* In addition to docstrings, these changes require updates to the actual documentation. Firstly there is the description of the YAML schema in `pyrobosim/docs/source/yaml/world_schema.rst` which should be updated. Secondly, I think we could benefit from having an actual new documentation page that shows how planners work, how to make a new one, etc.

ibrahiminfinite commented:
Ah all the demos I did not try running are the failing ones lol.
I did not try running all the demos since the **59** tests were passing in pytest.

sea-bass commented:
> Ah all the demos I did not try running are the failing ones lol. I did not try running all the demos since the **59** tests were passing in pytest.

Yeah... unfortunately the test coverage doesn't really handle any of the GUI and/or ROS stuff now, so I tend to manually run and test the "main 4" demos. These being:
1. `demo.py` (with / without YAML file and multirobot)
2. `demo_pddl.py` (with multiple planning domain)
3. `demo.launch.py` (the ROS version of 1.)
4. `demo_pddl.launch.py` (the ROS version of 2)

ibrahiminfinite commented:
> > Ah all the demos I did not try running are the failing ones lol. I did not try running all the demos since the **59** tests were passing in pytest.
> 
> Yeah... unfortunately the test coverage doesn't really handle any of the GUI and/or ROS stuff now, so I tend to manually run and test the "main 4" demos. These being:
> 
> 1. `demo.py` (with / without YAML file and multirobot)
> 2. `demo_pddl.py` (with multiple planning domains, there are 5 at the moment)
> 3. `demo.launch.py` (the ROS version of 1.)
> 4. `demo_pddl.launch.py` (the ROS version of 2)

For the pddl demos, are all 5 domains supposed to complete all steps successfully ?
Is it expected for it to sometimes fail due to situations like not being able to sample a free pose at the placement location ?

sea-bass commented:
> For the pddl demos, are all 5 domains supposed to complete all steps successfully ? Is it expected for it to sometimes fail due to situations like not being able to sample a free pose at the placement location ?

Ah yes, this is expected with the first 3 domains I think. Sorry, didn't specify.

ibrahiminfinite commented:
> > Ah all the demos I did not try running are the failing ones lol. I did not try running all the demos since the **59** tests were passing in pytest.
> 
> Yeah... unfortunately the test coverage doesn't really handle any of the GUI and/or ROS stuff now, so I tend to manually run and test the "main 4" demos. These being:
> 
> 1. `demo.py` (with / without YAML file and multirobot)
> 2. `demo_pddl.py` (with multiple planning domains, there are 5 at the moment)
> 3. `demo.launch.py` (the ROS version of 1.)
> 4. `demo_pddl.launch.py` (the ROS version of 2)

All of these demos are now running.

ibrahiminfinite commented:
The docstrings have been moved to the respective implementations instead of being in the `PathPlanner` class.

The major thing left now is the documentation for how to use and extend the planners.

ibrahiminfinite commented:
I have added the documentation page for path planner and also updated the `world_schema.rst`
The added page is here : https://pyrobosim--82.org.readthedocs.build/en/82/usage/path_planners.html

sea-bass commented on docs/source/usage/path_planners.rst:
```diff
@@ -0,0 +1,86 @@
+
+Note:
+
+- Planner implementations that need to display graphs should provide a `get_graphs` and set the `graphs` attribute of `PlannerBase` like in
+  :py:func:`pyrobosim.navigation.rrt.RRTPlannerPolygon.get_graphs` and :py:func:`pyrobosim.navigation.rrt.RRTPlanner.plan`
```
For this stuff, you can actually add note blocks: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html

Also,

```suggestion
- Planner implementations that need to display graphs should provide a `get_graphs()` method and set the `graphs` attribute of `PlannerBase` like in
  :py:func:`pyrobosim.navigation.rrt.RRTPlannerPolygon.get_graphs` and :py:func:`pyrobosim.navigation.rrt.RRTPlanner.plan`.
```

sea-bass commented on pyrobosim/examples/demo_astar.py:
```diff
@@ -1,39 +1,38 @@
 
-def test_astar_grid():
+def demo_astar():
+    """Creates an A* planner and plans a path using grid."""
```
Maybe say

```suggestion
    """Creates an occupancy grid based A* planner and plans a path."""
```

sea-bass commented on pyrobosim/pyrobosim/core/robot.py:
```diff
@@ -284,8 +286,10 @@ def place_object(self, pose=None):
 
         # Validate the robot location
         loc = self.location
+        if isinstance(loc, str):
+            loc = self.world.get_entity_by_name(self.location)
         if not isinstance(loc, ObjectSpawn):
-            warnings.warn("Not an object spawn. Cannot place object.")
+            warnings.warn(f"{loc.name} is not an object spawn. Cannot place object.")
```
Careful! If `loc` is something else (namely, `None`), this will error out... which I think was the main reason this wasn't in.

Maybe you can do an `elif ifinstance(loc, Location)` and then a catch-all `else` after?

ibrahiminfinite replied:
I think we can go with `loc` instead of `loc.name` and that should handle things well since the entities have  a `__repr__` and it will print `None` if it is `NoneType`

sea-bass commented on pyrobosim/pyrobosim/core/world.py:
```diff
@@ -1291,3 +1023,181 @@ def remove_robot(self, robot_name):
+        :type start: :class:`Pose`
+        :param goal: Goal node
+        :type goal: :class:`Pose`
+        :param ignore_max_dist: If True, ignores maximum connection distance.
+        :type ignore_max_dist: bool, optional
```
I think this param changed, so docstring should also change

sea-bass commented on pyrobosim/pyrobosim/gui/world_canvas.py:
```diff
@@ -308,19 +286,24 @@ def show_planner_and_path(self, robot=None):
-                self.axes, path_color=color
-            )
+            path_planner_artists = robot.path_planner.plot(self.axes, path_color=color)
+            if "graph" in path_planner_artists:
+                for artist in self.path_planner_artists["graph"]:
```
A "hack" here is:

```suggestion
            for artist in self.path_planner_artists.get("graph", []):
```

Same for the path artists that follow

ibrahiminfinite replied:
Oh this is an interesting pattern !



```python
self.path_planner_artists["graph"] = path_planner_artists["graph"]
```
but this line also depends on the check , so even if we use the pattern you suggested we will still need to have a key check.

ibrahiminfinite replied:
Was able to make use of this pattern somewhere else in the planners to solve a different problem.
Thanks !

sea-bass commented on pyrobosim/pyrobosim/gui/world_canvas.py:
```diff
@@ -418,22 +401,25 @@ def navigate(self, robot, goal):
+            robot.current_goal = goal
+            goal_pose = self.world.graph_node_from_entity(goal, robot=robot).pose
+            path = robot.plan_path(robot.pose, goal_pose)
+            print(f"path has : {path.num_poses} waypoints")
```
Remove this print statement?

sea-bass reviewed (commented):
Thanks for going through this! Besides the new docs, I didn't find all that much in the source code itself. Nice work.

Testing the examples, I did find a few things:

---

When I ran `demo.py` I moved the robot to `counter0` and tried to pick a water object, to which I got the following error:

```
[robot0] Picking water0
/home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/core/robot.py:230: UserWarning: water0 is at counter0_left and robot is at counter0. Cannot pick.
```

and same when I try place an object:

```
[robot0] Placing water0
/home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/core/robot.py:292: UserWarning: desk0 is not an object spawn. Cannot place object.
  warnings.warn(f"{loc.name} is not an object spawn. Cannot place object.")
```

So I think something broke in the robot's current location being the parent `Location` instead of the `ObjectSpawn`.

---

In the multirobot demo, it would be nice if the 3 robots had different path planners.

And when running with the test world file, I get this:

```
python pyrobosim/examples/demo.py --world-file test_world_multirobot.yaml

/home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/gui/world_canvas.py:305: UserWarning: Robot does not have a planner
  warnings.warn("Robot does not have a planner")
```

In fact, I looked at the `test_world_multirobot.yaml` file and this one still has a `global_path_planner` field. So I guess it needs to be updated.

---

The PDDLStream and ROS demos work great otherwise, so we're close!

sea-bass commented on pyrobosim/pyrobosim/data/test_world.yaml:
```diff
@@ -22,12 +22,15 @@ robots:
+      #   inflation_radius: 0.15
+      max_connection_dist: 0.25
+      bidirectional: true
+      rrt_star: false
```
Sorry, another comment -- can you make `rrt_star: true` in this file so the examples work the same as before?

ibrahiminfinite commented:
> When I ran `demo.py` I moved the robot to `counter0` and tried to pick a water object, to which I got the following error:
> 
> ```
> [robot0] Picking water0
> /home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/core/robot.py:230: UserWarning: water0 is at counter0_left and robot is at counter0. Cannot pick.
> ```
> 
> and same when I try place an object:
> 
> ```
> [robot0] Placing water0
> /home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/core/robot.py:292: UserWarning: desk0 is not an object spawn. Cannot place object.
>   warnings.warn(f"{loc.name} is not an object spawn. Cannot place object.")
> ```
> 
> So I think something broke in the robot's current location being the parent `Location` instead of the `ObjectSpawn`.

The robot is able to pick up the object when the navigation target is `water0`.

Is the robot expected to pick up or place  the object irrespective of if the object is on `counter0_left` or `counter0_right` when the robot is at `counter0` ?

ibrahiminfinite commented:
> In the multirobot demo, it would be nice if the 3 robots had different path planners. Conveniently, we can have one of each here :)
> 
> And when running with the test world file, I get this:
> 
> ```
> python pyrobosim/examples/demo.py --world-file test_world_multirobot.yaml
> 
> /home/sebastian/personal/pyrobosim_ws/src/pyrobosim/pyrobosim/pyrobosim/gui/world_canvas.py:305: UserWarning: Robot does not have a planner
>   warnings.warn("Robot does not have a planner")
> ```
> 
> In fact, I looked at the `test_world_multirobot.yaml` file and this one still has a `global_path_planner` field. So I guess it needs to be updated.

This has been fixed .

sea-bass commented:
> The robot is able to pick up the object when the navigation target is `water0`.

Great. It should also work if the target is just the object type (water) and if left blank, it should pick an object from that location at random.

> Is the robot expected to pick up or place the object irrespective of if the object is on `counter0_left` or `counter0_right` when the robot is at `counter0` ?

No, that should not work. Actually, the robot can't be at "counter0", which is the maik problem.

When you tell it to go to "counter0", it internally picks either the "counter0_left" or "counter0_right" location. So when you get the robot location after navigating, it should return one of those and the object resolution should "just work".

sea-bass commented:
> I am not sure if it is the docstring that is wrong or if the conversion should happen in `WorldCanvas`

I think the docstring is out of date. Both of these should resolve locations / objects all the way down to the specific instance.

ibrahiminfinite commented:
> When you tell it to go to "counter0", it internally picks either the "counter0_left" or "counter0_right" location. So when you get the robot location after navigating, it should return one of those and the object resolution should "just work".

In this case,  if the robot is at `counter0_left` but object is at `counter0_right`, is the robot  expected to be able to pick up the object ?

ibrahiminfinite commented:
``` python

        if (
            isinstance(entity, ObjectSpawn)
            or isinstance(entity, Room)
            or isinstance(entity, Hallway)
        ):
            graph_nodes = entity.graph_nodes
        elif isinstance(entity, Object):
            graph_nodes = entity.parent.graph_nodes
        elif isinstance(entity, Location):
            graph_nodes = entity.children[0].graph_nodes
            # TODO: Select a child node
        else:
            warnings.warn(f"Cannot get graph node from {entity}")
            return None
```

The `ObjectSpawn` selection for `Location` here look hardcoded to the first one, which results in the robot always going to `counter0_left` when navigating to `counter0`

Changing  it to select a random `ObjectSpawn` like this takes care of the issue
```python
   graph_nodes = entity.children[0].graph_nodes 
```
to
``` python 
graph_nodes = random.choice(entity.children).graph_nodes
 ```

sea-bass commented:
> ```python
>         if (
>             isinstance(entity, ObjectSpawn)
>             or isinstance(entity, Room)
>             or isinstance(entity, Hallway)
>         ):
>             graph_nodes = entity.graph_nodes
>         elif isinstance(entity, Object):
>             graph_nodes = entity.parent.graph_nodes
>         elif isinstance(entity, Location):
>             graph_nodes = entity.children[0].graph_nodes
>             # TODO: Select a child node
>         else:
>             warnings.warn(f"Cannot get graph node from {entity}")
>             return None
> ```
> 
> The `ObjectSpawn` selection for `Location` here look hardcoded to the first one, which results in the robot always going to `counter0_left` when navigating to `counter0`
> 
> Changing it to select a random `ObjectSpawn` like this takes care of the issue
> 
> ```python
>    graph_nodes = entity.children[0].graph_nodes 
> ```
> 
> to
> 
> ```python
> graph_nodes = random.choice(entity.children).graph_nodes
> ```
> 
> This way you can move from one spawn location to another by trying to navigate to the same parent location.

Where is that code? Better to link it than copy-paste it.

Anyways, choosing the first vs. a random location won't fully address the issue, right? I think the problem is that whatever is grabbing `robot.location` is getting the parent `Location` and not the child `ObjectSpawn`. Even if you always picked the first spawn in a location, the information you get back should agree with where the robot actually is?

ibrahiminfinite commented:
> Anyways, choosing the first vs. a random location won't fully address the issue, right? I think the problem is that whatever is grabbing `robot.location` is getting the parent `Location` and not the child `ObjectSpawn`. Even if you always picked the first spawn in a location, the information you get back should agree with where the robot actually is?

The random selection mostly helps when you are using the GUI and the object you want to grab is at a different  `ObjectSpawn` that the `first` `ObjectSpawn`.

The code is in [graph_node_from_entity()](https://github.com/sea-bass/pyrobosim/blob/8bd74e694d88223c62fce5d2b50349332074aa64/pyrobosim/pyrobosim/core/world.py#L1145)

sea-bass commented:
> The random selection mostly helps when you are using the GUI and the object you want to grab is at a different `ObjectSpawn` than the `first` `ObjectSpawn`.

The input to that text box is a query, so you could actually type in e.g. "counter water" or "water0" and the right object spawn should be automatically resolved. But anyways, in the event that the query has multiple possible answers, I have no strong preference on choosing the first spawn vs. a random one. It's just as arbitrary.

sea-bass commented on pyrobosim/pyrobosim/gui/world_canvas.py:
```diff
@@ -402,9 +402,9 @@ def navigate(self, robot, goal):
-            path = robot.plan_path(robot.pose, goal_pose)
+            goal_node = self.world.graph_node_from_entity(goal, robot=robot)
+            robot.current_goal = goal_node.parent
+            path = robot.plan_path(robot.pose, goal_node.pose)
```
I know this was here before, but I wonder if it's more principled to use [`resolve_to_location()`](https://github.com/sea-bass/pyrobosim/blob/8bd74e694d88223c62fce5d2b50349332074aa64/pyrobosim/pyrobosim/utils/knowledge.py#L170) here instead.

It may prevent you from having to get a graph node just to unpack it.

ibrahiminfinite replied:
The `resolve_to_location()` methods returns

```
    :return: The location or object spawn that meets the category and/or room filters, or None.
    :rtype: :class:`pyrobosim.core.locations.Location`/:class:`pyrobosim.core.locations.ObjectSpawn`
```

But goal is already an `Entity` since it is converted in `PyRoboSimMainWindow.on_navigate_click()`
The reason I used resolve to `graph_node_from_entity()` was because that is how it is done in the [`World.find_path()`](https://github.com/sea-bass/pyrobosim/blob/d4340278d21a003bcaadedda1f680bcaf96d7123/pyrobosim/pyrobosim/core/world.py#L760) in the  [ `WorldCanvas.navigate()` ](https://github.com/sea-bass/pyrobosim/blob/d4340278d21a003bcaadedda1f680bcaf96d7123/pyrobosim/pyrobosim/gui/world_canvas.py#L422) in `main`

sea-bass commented on pyrobosim/examples/demo.py:
```diff
@@ -10,8 +10,9 @@
-from pyrobosim.navigation import ConstantVelocityExecutor, RRTPlanner
+from pyrobosim.navigation import ConstantVelocityExecutor, PathPlanner
 from pyrobosim.utils.general import get_data_folder
+from pyrobosim.navigation import occupancy_grid_from_world
```
Make sure it's in alphabetical order, just push this one line up I think

sea-bass commented on pyrobosim/examples/demo.py:
```diff
@@ -97,13 +98,24 @@ def create_world(multirobot=False):
-        name="robot",
+
+    robot0 = Robot(
+        name="robot0",
```
I guess I forgot to specify, but if you don't give the robot a name, the `World.add_robot()` method should automatically rename them in this same fashion: `robot0`, `robot1`, etc... 

Up to you if you wanna keep these lines for the 3 robots here, no big difference.

ibrahiminfinite replied:
I was aware of the automatic naming, I just chose to name them explicitly.

sea-bass commented on pyrobosim/pyrobosim/gui/world_canvas.py:
```diff
@@ -308,19 +286,24 @@ def show_planner_and_path(self, robot=None):
+            path_planner_artists = robot.path_planner.plot(self.axes, path_color=color)
+            if "graph" in path_planner_artists:
+                for artist in self.path_planner_artists["graph"]:
+                    artist.remove()
```
I think the reason switching between robots might be behaving unpredictably is that the artists should be removed always, regardless of whether the new thing to plot has `graph` or `path` defined.

Notice in lines 311-312 deleted, we were always removing everything first, and then adding if new things existed.

Same for the `artist.remove()` call in the path planner artists below.

sea-bass commented on pyrobosim_ros/examples/demo.py:
```diff
@@ -15,6 +15,7 @@
 from pyrobosim.utils.general import get_data_folder
 from pyrobosim.utils.pose import Pose
 from pyrobosim_ros.ros_interface import WorldROSWrapper
+from pyrobosim.navigation import PathPlanner
```
Same here, move this in alphabetical order

sea-bass commented on pyrobosim/examples/demo.py:
```diff
@@ -97,13 +98,24 @@ def create_world(multirobot=False):
         grasp_generator=GraspGenerator(grasp_props),
     )
-    world.add_robot(robot, loc="kitchen")
+    planner_config_astar = {
+        "grid": occupancy_grid_from_world(
+            world, resolution=0.05, inflation_radius=1.5 * robot0.radius
+        ),
+        "diagonal_motion": True,
+        "heuristic": "euclidean",
+    }
```
The planner in this demo is A*, but in the YAML file and in the ROS demo it's RRT.

Maybe switch this to also default to RRT with the same settings. But see my bug with the A* path yaws in the main review!

sea-bass commented on test/planning/test_pddlstream_nav.py:
```diff
@@ -59,8 +59,14 @@ def create_test_world(add_hallway=True):
     # Create a search graph and motion planner
-    world.create_search_graph(max_edge_dist=3.0, collision_check_dist=0.05)
-    rrt = RRTPlanner(world, bidirectional=True, rrt_star=True)
+    # world.create_search_graph(max_edge_dist=3.0, collision_check_dist=0.05)
```
remove this commented-out line

sea-bass reviewed (commented):
I just tested again and found some more bugs.

First, I realized that the robot orientation doesn't actually change with the A* paths. Maybe the `fill_yaws()` call got lost?

![image](https://user-images.githubusercontent.com/4603398/230132582-9b3504bd-1d3e-4cb1-9577-6a18458976f1.png)

Also, if the robot is at `counter0_right` and you hit "Pick", it seems we're still looking for objects in `counter0_left`, which seems to be a bug.

![image](https://user-images.githubusercontent.com/4603398/230132865-86e814ea-3007-42e5-a4ea-bd064bfdf42c.png)

When I start the multirobot demo, navigating with the PRM is animating so slow it never re-renders. Are the graph artists getting re-displayed constantly or something?

![image](https://user-images.githubusercontent.com/4603398/230133515-d3136b22-58af-4b5d-a6e8-055a24b5ceb9.png)

Also in the multirobot demo, when I switch from the PRM robot `robot1` to the RRT robot `robot2` and then switch back out, the graph doesn't get cleared and the old RRT keeps showing up forever. I think the graph needs to get cleared out when robots/planners are switched. Added a comment in the UI code which I think will fix that.

sea-bass commented:
> Or do you mean if it is at `counter0_right` and we hit `Pick` , it should only look for objects in `counter0_right`?

Exactly this!

ibrahiminfinite commented:
I think the reason for high rendering times is because every edge is getting added twice now.
Edge(A, B) and Edge(B, A) while these are conceptually same, the current logic does not check if one or the other exist and restrict a redundant adding of edges.
https://github.com/sea-bass/pyrobosim/blob/a7afed2776cbafc57aac3c556c4d17a6187142f5/pyrobosim/pyrobosim/navigation/world_graph.py#L75

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/world_graph.py:
```diff
@@ -0,0 +1,128 @@
+                self.edges.discard(edge)
+        self.update_neighbors()
+
+    def update_neighbors(self):
```
This function can be optimized to avoid looping through all the edges.

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/prm.py:
```diff
@@ -41,18 +46,28 @@ def reset(self):
+            if self.is_connectable(node.pose, other.pose):
+                self.graph.add_edge(node, other)
+
     def plan(self, start, goal):
```
During the planning phase of PRM in `plan()` method, the 'start` and `goal` nodes and edges from them should be removed after the planning is done.
Otherwise the graph keeps growing.

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/rrt.py:
```diff
@@ -261,22 +276,20 @@ def rewire_node(self, graph, n_tgt):
             dist = n.pose.get_linear_distance(n_tgt.pose)
             if (n != n_tgt) and (dist <= self.rewire_radius):
```
Comparison here

sea-bass replied:
Based on e.g. [this](https://www.hylkerozema.nl/2021/05/10/object-oriented-programming-how-to-compare-arguments-of-methods/), the default python behavior is to compare object references. So this will return true when the variables point to the same Node object.

Overloading __eq__ would make it, e.g., so that two different Node objects with similar poses would also come out as equal, which may be helpful, but not necessary in this case?

ibrahiminfinite replied:
> which may be helpful, but not necessary in this case?

Looks like this is enough.
The issue is with the new `Graph` implementation. Will see if I can change it to use the default python mechanism instead.

ibrahiminfinite commented on pyrobosim/pyrobosim/navigation/rrt.py:
```diff
@@ -261,22 +276,20 @@ def rewire_node(self, graph, n_tgt):
         if n_rewire is not None:
             n_tgt.parent = n_rewire
             for e in graph.edges:
-                if e.n0 == n_tgt or e.n1 == n_tgt:
```
And here

ibrahiminfinite reviewed (commented):
This was already there , but I don't think these `Node` object comparisons actually work  since there is no `__eq__`  or `__hash__` operator defined.
Even if 2 nodes had the same attributes the nodes would be reported as not equal.
Is there something I am missing here ? @sea-bass

sea-bass commented on pyrobosim/pyrobosim/navigation/world_graph.py:
```diff
@@ -70,8 +70,17 @@ def remove_node(self, node):
+            other.neighbors.discard(node)
         self.nodes.discard(node)
 
+        edges_to_remove = []
+        for edge in self.edges:
+            if edge.nodeA == node or edge.nodeB == node:
+                edges_to_remove.append(edge)
+        for edge in edges_to_remove:
+            self.edges.discard(edge)
```
Why remove the `remove_edge` function and move it in here, rather than directly call it and keep the ability to remove an edge with a single function?

sea-bass replied:
I get it now. Will re-add the `remove_edge()` method regardless but keep your implementation.

sea-bass commented on pyrobosim/pyrobosim/navigation/prm.py:
```diff
@@ -27,12 +32,12 @@ def __init__(self, world, max_nodes=100, max_connection_dist=2.0):
+        self.compress_path = compress_path
+
+        self.path_finder = AStarGraph()
+        # for key, value in planner_config.items():
+        #     setattr(self, key, value)
```
Remove? Or uncomment?

sea-bass reviewed (commented):
I dug into that issue about the counter and object resolution and am going to say this is working as intended, so no need to look at that further. It's fine, and not a change introduced by your updates at all.

In testing, I found an issue that the A* planner always started and ended with zero yaw because the actual start/goal orientations were fully discarded. So I fixed that plus some other minor stuff.

I just have 1 last major comment about removing the `remove_edge()` function in the graph class. Once that is resolved, should be good.

sea-bass reviewed (approved):
I went ahead and fixed all the remaining things I found, so LGTM.

@ibrahiminfinite let me know if you maybe wanna do your own round of final testing and then we can get this in.

Thanks for the work on this one! Big improvements to planners, for sure.

ibrahiminfinite commented:
> It's fine, and not a change introduced by your updates at all.

I was wondering what unit tests I should put in to make sure this was not breaking anything.
SInce the issue was not introduced from this PR.
We can get this in and add unit tests for small utility functions like `entity_from_query` etc as the next step.

Thanks for guiding me !!
